src/app/clusters/joint-fabric-administrator-server/joint-fabric-administrator-server.cpp

```diff
@@ -102,31 +100,82 @@ void MatterJointFabricAdministratorPluginServerInitCallback()
     const chip::app::Clusters::JointFabricAdministrator::Commands::ICACCSRRequest::DecodableType & commandData)
 {
     MATTER_TRACE_SCOPE("ICACCSRRequest", "JointFabricAdministrator");
+    ChipLogProgress(Zcl, "JointFabricAdministrator: Received an ICACCSRRequest command");
+
+    auto nonDefaultStatus           = Status::Success;
+    const FabricInfo * fabricInfo   = RetrieveCurrentFabric(commandObj);
+    auto & failSafeContext          = Server::GetInstance().GetFailSafeContext();
+    auto & jointFabricAdministrator = Server::GetInstance().GetJointFabricAdministrator();
+
+    uint8_t buf[Credentials::kMaxDERCertLength];
+    MutableByteSpan icacCsr(buf, Credentials::kMaxDERCertLength);
+    Commands::ICACCSRResponse::Type response;
+
+    // If current fabric is not available, command was invoked over PASE which is not legal
+    VerifyOrExit(fabricInfo != nullptr, nonDefaultStatus = Status::InvalidCommand);
 
-    ChipLogProgress(Zcl, "JointFabricAdministrator: Received a ICACCSRRequest command");
+    VerifyOrExit(failSafeContext.IsFailSafeArmed(commandObj->GetAccessingFabricIndex()),
+                 nonDefaultStatus = Status::FailsafeRequired);
+
+    /* TODO spec.: If the <<ref_FabricTableVendorIdVerificationProcedure, FabricFabric Table Vendor ID Verification Procedure>>
+     * has not been executed against the initiator of this command, the command SHALL fail
+     * with a <<ref_JFVidNotVerified, JfVidNotVerified>> status code SHALL be sent back to the initiator.*/
+
+    VerifyOrExit(jointFabricAdministrator.GetDelegate() != nullptr, nonDefaultStatus = Status::Failure);
+    VerifyOrExit(jointFabricAdministrator.GetDelegate()->GetIcacCsr(icacCsr) == CHIP_NO_ERROR, nonDefaultStatus = Status::Failure);
+
+    response.icaccsr = icacCsr;
+    commandObj->AddResponse(commandPath, response);
+
+exit:
+    if (nonDefaultStatus != Status::Success)
+    {
+        commandObj->AddStatus(commandPath, nonDefaultStatus);
+    }
 
     return true;
 }
 
-// TODO
 bool emberAfJointFabricAdministratorClusterAddICACCallback(
     chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
     const chip::app::Clusters::JointFabricAdministrator::Commands::AddICAC::DecodableType & commandData)
 {
     MATTER_TRACE_SCOPE("AddICAC", "JointFabricAdministrator");
+    ChipLogProgress(Zcl, "JointFabricAdministrator: Received an AddICAC command");
+
+    auto nonDefaultStatus         = Status::Success;
+    const FabricInfo * fabricInfo = RetrieveCurrentFabric(commandObj);
+    auto & failSafeContext        = Server::GetInstance().GetFailSafeContext();
 
-    ChipLogProgress(Zcl, "JointFabricAdministrator: Received a AddICAC command");
+    // If current fabric is not available, command was invoked over PASE which is not legal
+    VerifyOrExit(fabricInfo != nullptr, nonDefaultStatus = Status::InvalidCommand);
 
+    VerifyOrExit(failSafeContext.IsFailSafeArmed(commandObj->GetAccessingFabricIndex()),
```

Comment:
The spec has the fail-safe check before the CASE check, no?

Reply:
Right, fixed this.

Reply:
Spec lists the checks in order, but looks like the outstanding spec PR (not linked in this PR's description, sadly) changes the order... So ok.